doc/specs/stdlib_hash_functions.md

```diff
@@ -0,0 +1,1744 @@
+
+#####  Class
+
+Pure function
+
+##### Arguments
+
+`key`: shall be a scalar of type default character expression or a
+Rank 1 integer vector expression of kind `INt8`, `INT16`, `INT32`, or
+`INTT64`. It is an `intent(in)` argument.
+
+`seed`: shall be a two element integer vector expression of kind
+`INT64`. It is an `intent(in)` argument.
+
+##### Result
+
+The result is a two element integer vector of kind `INT64`.
+
+##### Note
+
+`SPOOKY_HASH` is an implementation of the 64 bit version 2 of
+SpookyHash of Bob Jenkins. The code was designed for Little-Endian
+compilers. The output is different on Big Endian compilers, but still
+probably as good quality. It is often used as a 64 bit hash using the
+first element of the returned value, but can be used as a 128 bit
+hash. This version of `SPOOKY_HASH` has good performance on small keys
+and excellent performance on long keys. It passes all the SMHasher tests
+and has no known bad seeds.
+
+##### Example
+
+```fortran
+    program demo_spooky_hash
+      use stdlib_64_bit_hash_codes, only: new_spooky_hash_seed, &
+          spooky_hash
+      use iso_fortran_env, only: int64 
+      implicit none
+      integer, allocatable :: key(:)
+      integer(int64) :: hash(2), seed(2), source
+      key = [ 0_int64, 1_int64, 2_int64, 3_int64 ]
+      seed = [ 119_int64, 2_int64**41-1 ]
+      call new_spooky_hash_seed( seed )
+      hash = spooky_hash( key, seed )
+      print *, seed, hash
+    end program demo_spooky_hash
+```
+
+#### `UNIVERSAL_MULT_HASH` - maps an integer to a smaller number of bits
+
+##### Status
+
+Experimental
+
+##### Description
+
+Calculates an `nbits` hash code from a 64 bit integer. This is useful
+in mapping a hash value to a range 0 to `2**nbits-1`.
+
+##### Syntax
+
+`code = [[stdlib_64_bit_hash_codes:universal_mult_hash]]( key, seed, nbits )`
+
+##### Class
+
+Pure function
+
+##### Arguments
+
+`key`: Shall be an integer of kind `INT64`. It is an `intent(in)`
+argument.
+
+`seed`: Shall be an integer of kind `INT64`. It is an `intent(in)`
+argument. It should be an odd value.
+
+`nbits` Shall be a default integer with `0 < nbits < 64`. It is an
+`intent(in)` argument.  It must be an odd integer.
+
+##### Result
+
+The result is an integer of kind `INT64` with at most the lowest
+`nbits` nonzero.
+
+##### Note
+
+`UNIVERSAL_MULT_HASH` is an implementation of the Universal
+Multiplicative Hash of M. Dietzfelbinger, et al.
+It multiplies the `KEY` by `SEED`, and returns the
+`NBITS` upper bits of the product as the lowest bits of the result.
+
+##### Example
+
+
+```fortran
+    program demo_universal_mult_hash
+      use stdlib_32_bit_hash_codes, only: odd_random_integer, &
+          universal_mult_hash
+      use iso_fortran_env, only: int64
+      implicit none
+      integer, allocatable :: array1(:)
+      integer(int64) :: hash, i, seed, source
+      seed = 0
+      allocate( array1(0:2**6-1) )
+      do i = 0, 2**6-1
+          array(i) = i
```

Comment:
Not sure to understand the usefullness of this part of the code in this example

Reply:
I have changed the do loop to 
        array1 = 0

The rest of the code then inserts 42 into the array nat its hash value. so the location of the 42 reveals its hash value.